File: source/math/float4.cpp
#include "math/float4.h"

#include <iostream>

#include "utility/convert.h"


// Constructors
kl::float4::float4()
{
	x = 0;
	y = 0;
	z = 0;
	w = 0;
}
kl::float4::float4(float a)
{
	x = a;
	y = a;
	z = a;
	w = a;
}
kl::float4::float4(float x, float y, float z, float w)
{
	this->x = x;
	this->y = y;
	this->z = z;
	this->w = w;
}
kl::float4::float4(const kl::int4& v)
{
	x = float(v.x);
	y = float(v.y);
	z = float(v.z);
	w = float(v.w);
}
kl::float4::float4(const kl::float2& v, float z, float w)
{
	x = v.x;
	y = v.y;
	this->z = z;
	this->w = w;
}
kl::float4::float4(float x, const kl::float2& v, float w)
{
	this->x = x;
	y = v.x;
	z = v.y;
	this->w = w;
}
kl::float4::float4(float x, float y, const kl::float2& v)
{
	this->x = x;
	this->y = y;
	z = v.x;
	w = v.y;
}
kl::float4::float4(const kl::float2& v1, const kl::float2& v2)
{
	x = v1.x;
	y = v1.y;
	z = v2.x;
	w = v2.y;
}
kl::float4::float4(const kl::float3& v)
{
	x = v.x;
	y = v.y;
	z = v.z;
	w = 1;
}
kl::float4::float4(const kl::float3& v, float w)
{
	x = v.x;
	y = v.y;
	z = v.z;
	this->w = w;
}
kl::float4::float4(float x, const kl::float3& v)
{
	this->x = x;
	y = v.x;
	z = v.y;
	w = v.z;
}
kl::float4::float4(const kl::color& color)
{
	static const float toFloatCol = 1.0f / 255;
	x = color.r * toFloatCol;
	y = color.g * toFloatCol;
	z = color.b * toFloatCol;
	w = color.a * toFloatCol;
}

// Getter
kl::float3 kl::float4::xyz() const
{
	return kl::float3(x, y, z);
}

// Addition
kl::float4 kl::float4::add(const kl::float4& obj) const
{
	return kl::float4(x + obj.x, y + obj.y, z + obj.z, w + obj.w);
}
kl::float4 kl::float4::operator+(const kl::float4& obj) const
{
	return this->add(obj);
}
void kl::float4::operator+=(const kl::float4& obj)
{
	x += obj.x; y += obj.y; z += obj.z; w += obj.w;
}

// Subtraction
kl::float4 kl::float4::sub(const kl::float4& obj) const
{
	return kl::float4(x - obj.x, y - obj.y, z - obj.z, w - obj.w);
}
kl::float4 kl::float4::operator-(const kl::float4& obj) const
{
	return this->sub(obj);
}
void kl::float4::operator-=(const kl::float4& obj)
{
	x -= obj.x; y -= obj.y; z -= obj.z; w -= obj.w;
}

// Multiplication
kl::float4 kl::float4::mul(float a) const
{
	return kl::float4(x * a, y * a, z * a, w * a);
}
kl::float4 kl::float4::operator*(float a) const
{
	return this->mul(a);
}
void kl::float4::operator*=(float a)
{
	x *= a; y *= a; z *= a; w *= a;
}
kl::float4 kl::float4::mul(const kl::float4& obj) const
{
	return kl::float4(x * obj.x, y * obj.y, z * obj.z, w * obj.w);
}
kl::float4 kl::float4::operator*(const kl::float4& obj) const
{
	return this->mul(obj);
}
void kl::float4::operator*=(const kl::float4& obj)
{
	x *= obj.x; y *= obj.y; z *= obj.z; w *= obj.w;
}

// Division
kl::float4 kl::float4::div(float a) const
{
	return this->mul(1 / a);
}
kl::float4 kl::float4::operator/(float a) const
{
	return this->div(a);
}
void kl::float4::operator/=(float a)
{
	this->operator*=(1 / a);
}
kl::float4 kl::float4::div(const kl::float4& obj) const
{
	return kl::float4(x / obj.x, y / obj.y, z / obj.z, w / obj.w);
}
kl::float4 kl::float4::operator/(const kl::float4& obj) const
{
	return this->div(obj);
}
void kl::float4::operator/=(const kl::float4& obj)
{
	x /= obj.x; y /= obj.y; z /= obj.z; w /= obj.w;
}

// Comparison
bool kl::float4::equals(const kl::float4& obj) const
{
	return x == obj.x && y == obj.y && z == obj.z && w == obj.w;
}
bool kl::float4::operator==(const kl::float4& obj) const
{
	return this->equals(obj);
}
bool kl::float4::operator!=(const kl::float4& obj) const
{
	return !this->equals(obj);
}

// Returns a negated vector
kl::float4 kl::float4::negate() const
{
	return this->mul(-1);
}

// Returns an absolute vector
kl::float4 kl::float4::abso() const
{
	return kl::float4(abs(x), abs(y), abs(z), abs(w));
}

// Returns the vectors length
float kl::float4::length() const
{
	return sqrt(x * x + y * y + z * z + w * w);
}

// Retruns a normalized vector
kl::float4 kl::float4::normalize() const
{
	return this->div(this->length());
}

// Returns the dot product
float kl::float4::dot(const kl::float4& a) const
{
	return x * a.x + y * a.y + z * a.z + w * a.w;
}

// Returns the angle between the given vector and self
float kl::float4::angle(const kl::float4& a) const
{
	return kl::convert::toDegrees(acos(this->normalize().dot(a.normalize())));
}

// Overloading std::cout
std::ostream& kl::operator<<(std::ostream& os, const kl::float4& obj)
{
	os << "(" << kl::convert::toString(obj.x) << ", " << kl::convert::toString(obj.y) << ", " << kl::convert::toString(obj.z) << ", " << kl::convert::toString(obj.w) << ")";
	return os;
}

File: source/kl/math/Float4.java
package kl.math;

import kl.graphics.Color;

public class Float4 {
    public float x, y, z, w;

    public Float4() {
        x = 0.0f;
        y = 0.0f;
        z = 0.0f;
        w = 0.0f;
    }
    public Float4(float a) {
        x = a;
        y = a;
        z = a;
        w = a;
    }
    public Float4(float x, float y, float z, float w) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.w = w;
    }
    public Float4(Int4 v) {
        x = (float)v.x;
        y = (float)v.y;
        z = (float)v.z;
        w = (float)v.w;
    }
    public Float4(Float2 v, float z, float w) {
        x = v.x;
        y = v.y;
        this.z = z;
        this.w = w;
    }
    public Float4(float x, Float2 v, float w) {
        this.x = x;
        y = v.x;
        z = v.y;
        this.w = w;
    }
    public Float4(float x, float y, Float2 v) {
        this.x = x;
        this.y = y;
        z = v.x;
        w = v.y;
    }
    public Float4(Float2 v1, Float2 v2) {
        x = v1.x;
        y = v1.y;
        z = v2.x;
        w = v2.y;
    }
    public Float4(Float3 v, float w) {
        x = v.x;
        y = v.y;
        z = v.z;
        this.w = w;
    }
    public Float4(float x, Float3 v) {
        this.x = x;
        y = v.x;
        z = v.y;
        w = v.z;
    }
    public Float4(Float4 v) {
        x = v.x;
        y = v.y;
        z = v.z;
        w = v.w;
    }
    private static final float fromColor = 1.0f / 255.0f;
    public Float4(Color color) {
        x = color.getRed() * fromColor;
        y = color.getGreen() * fromColor;
        z = color.getBlue() * fromColor;
        w = color.getAlpha() * fromColor;
    }

    public Float2 xy() {
        return new Float2(x, y);
    }
    public Float3 xyz() {
        return new Float3(x, y, z);
    }

    public Float4 add(Float4 v) {
        return new Float4(x + v.x, y + v.y, z + v.z, w + v.w);
    }

    public Float4 subtract(Float4 v) {
        return new Float4(x - v.x, y - v.y, z - v.z, w - v.w);
    }

    public Float4 multiply(float a) {
        return new Float4(x * a, y * a, z * a, w * a);
    }
    public Float4 multiply(Float4 v) {
        return new Float4(x * v.x, y * v.y, z * v.z, w * v.w);
    }

    public Float4 divide(float a) {
        return multiply(1.0f / a);
    }
    public Float4 divide(Float4 v) {
        return new Float4(x / v.x, y / v.y, z / v.z, w / v.w);
    }

    public boolean equals(Float4 v) {
        return x == v.x && y == v.y && z == v.z && w == v.w;
    }

    public Float4 negate() {
        return multiply(-1.0f);
    }

    public float length() {
        return (float) Math.sqrt(x * x + y * y + z * z + w * w);
    }

    public Float4 normalize() {
        return divide(length());
    }

    float dot(Float4 a) {
        return x * a.x + y * a.y + z * a.z + w * a.w;
    }

    float angle(Float4 a) {
        return (float) Math.toDegrees(Math.acos(normalize().dot(a.normalize())));
    }

    public static Float4 getPosX() {
        return new Float4(1.0f, 0.0f, 0.0f, 0.0f);
    }
    public static Float4 getNegX() {
        return new Float4(-1.0f, 0.0f, 0.0f, 0.0f);
    }
    public static Float4 getPosY() {
        return new Float4(0.0f, 1.0f, 0.0f, 0.0f);
    }
    public static Float4 getNegY() {
        return new Float4(0.0f, -1.0f, 0.0f, 0.0f);
    }
    public static Float4 getPosZ() {
        return new Float4(0.0f, 0.0f, 1.0f, 0.0f);
    }
    public static Float4 getNegZ() {
        return new Float4(0.0f, 0.0f, -1.0f, 0.0f);
    }
    public static Float4 getPosW() {
        return new Float4(0.0f, 0.0f, 0.0f, 1.0f);
    }
    public static Float4 getNegW() {
        return new Float4(0.0f, 0.0f, 0.0f, -1.0f);
    }

    public String toString() {
        return "(" + x + ", " + y + ", " + z + ", " + w + ")";
    }
}
